docs/rfcs/0005-ecommerce-rest-api.md

```diff
@@ -0,0 +1,182 @@
+
+#### GET
+
+A response would look like
+
+    {
+        "items": [
+            {
+                "id": 3,  // this is ProductVersion.id, not Product.id
+                "description": "Supply Chain Management",
+                "price": "67.89"
+            }
+        ],
+        "coupons": [
+            {
+                "code": "DISCOUNT_15_OFF",
+                "amount": "0.15",
+                "targets": [3]  // ProductVersion.id. There may be more but products not already in cart will be filtered out.
+            }
+        ]
+    }
+
+An empty basket would return:
+
+    {
+        "items": [],
+        "coupons": []
+    }
+
+#### PATCH
+
+A user can PATCH the basket to update the items list or to attempt to use a coupon. Each
+time a user PATCHes this endpoint, the basket will be validated. If the basket is not valid
+a 400 error will be returned and the basket will be left unchanged.
+
+Also note that a basket passing validation is not a guarantee that the checkout will
+pass validation. This could happen if the coupon code was valid when it was being
+added to the basket but it was not valid by the time of checkout. A product may also
+have its validity changed, preventing checkout from working.
+
+TODO: what happens if a basket becomes invalid over time? I think this is somewhat rare
+but we should figure out how to handle it. We can just clear the basket but I don't know when
+we should do this validation.
+
+##### Clear a basket
+
+A user could clear a basket by patching:
+
+    {
+        "items": [],
+        "coupons": []
+    }
+    
+They could clear just the items in the basket:
+
+    {
+        "items": []
+    }
+    
+However this would leave the coupon code as is. If there was a coupon code it would
+fail validation because there is nothing to apply the coupon to, and the basket would
+be left unchanged. So there is no use case for this.
+
+A user could clear only the coupon:
+
+    {
+        "coupons": []
+    }
+    
+
+##### Add item to basket    
+A user could update the basket to have one item by patching:
+
+    {
+        "items": [
+            {
+                "id": 8   // this is ProductVersion.id, not Product.id
+            }
+        ]
+    }
+    
+Here the coupon is not one of the keys so it would be left as is. However the view
+will revalidate the coupon to make sure it can be applied to the new item list. If this
+validation fails a 400 response is returned and the basket is left unchanged.
+
+Also note that there is no way to incrementally add items to a basket, leaving existing
+items unchanged. The frontend will need to send the complete new list with `id` attributes
+for each product. However in practice we will not have more than one item in a basket.
+
+##### Add coupon to basket
+
+A user could apply a coupon to a basket:
+
+    {
+        "coupons": [
+            {
+                "code": "DISCOUNT_30_OFF"
+            }
+        ]
+    } 
```

Comment:
Seems like the `api/basket` API will handle all coupon-related functions including validation, which makes sense, so there's probably no need for #71 as a separate issue anymore?

Reply:
Hmm I think you're right

Reply:
If you'd like to split up the implementation of this differently (`api/basket` vs `api/products` vs `api/checkout` etc), let me know what part(s) you'd prefer me to work on.